### ISSUE TITLE
Links not being added to work items

### ISSUE THREAD

**Issue by davidlbueno** [solidify/jira-azuredevops-migrator#52]
When attempting to import work items into DevOps, some link types are not getting imported properly, specifically 'Related' and 'Duplicate' link types. 

Only some of the 'Hierarchy-Reverse' link types are getting imported properly, and none of the 'Related' and 'Duplicate' link types. Instead, a comment is added to the Work Item that looks like this:

```
Added link(s): [Added] EN-22/-1->EN-1458/-1 [System.LinkTypes.Hierarchy-Reverse];[Added] EN-22/-1->EN-24/-1 [System.LinkTypes.Relates]
```

The import log shows the following for that work item import:

```
[W][14:59:44] '[Added] EN-22/-1->EN-24/-1 [System.LinkTypes.Related]' - target work item for Jira 'EN-24' is not yet created in Azure DevOps/TFS.
```

The export file for this work item has the following link data:

```
"Links":[{"Change":0,"TargetOriginId":"EN-1458","TargetWiId":0,"WiType":"System.LinkTypes.Hierarchy-Reverse"},{"Change":0,"TargetOriginId":"EN-24","TargetWiId":0,"WiType":"System.LinkTypes.Related"}]
```

It seems that the links fail validation because their target doesn't exist yet, so they aren't added as links and comments with the link info are added to the work item instead. 

If that's the case, it seems like a chicken or the egg problem (the target work item must exist in order to import a link, but the import must have completed for all targets to exist). How would it be possible to import work items that reference each other without running into this problem?

Also, it seems odd that some of the 'Hierarchy" Linktypes are successfully imported, but none of the 'Related' or 'Duplicate' linktypes are. Are they handled differently?

Thanks in advance to all contributors for your work on this useful migration tool, any help would be greatly appreciated.

David

**Comment by MOlausson** [comment]
Hi @davidlbueno and thanks for bringing this to our attention. Links in Azure DevOps are bi-directional so if for instance the parent work item cannot create a child link since the child work item does not exist, the link should be established later when the child creates a parent link. But of course if the child item isn't in the import set than it will result in missing links. We'll investigate.

**Comment by davidlbueno** [comment]
Thanks for the response @MOlausson, Parent and child links of the Hierarchy-Reverse linkType appear to be working just as you described. Links that are of the Dupicate or Related Linktypes are not getting created on either end of the link however. For example, we have:

**Jira ID: EN-1934**
Export file excerpt:
```
"Links":[{"Change":0,"TargetOriginId":"EN-2027","TargetWiId":0,"WiType":"System.LinkTypes.Duplicate-Forward"}]
```
Excerpt from import log:
```
[W][11:29:18] '[Added] EN-1934/-1->EN-2027/-1 [System.LinkTypes.Duplicate-Forward]' - target work item for Jira 'EN-2027' is not yet created in Azure DevOps/TFS.
```
(The above message is added as a comment, but the link is not created)

**Jira ID: 2027**
Export file excerpt:
```
"Links":[]
```

It looks like the problem is that the export file for the 'copied to' (duplicate) item (EN2027) does not contain a link to the original (copied from) item (EN-1934). Do you have any suggestions as to why this may be happening? Are we using the wrong LinkType perhaps?

**Comment by madkoo** [comment]
Hi @davidlbueno, is this issue still applicable for you? 
I tried to repo your described steps using the same config 
` {
        "source": "Cloners",
        "target": "System.LinkTypes.Duplicate-Forward"
      }`
and setting up a duplicate issue in Jira. I was able to migrate those linked items successfully. Meaning  - 
1) Duplicate links and comments were added to the target links.  
2) Duplicate links were added to the origin items with comments.
